src/Components/Authentication/Mobile/LoginForm.tsx

```diff
@@ -22,15 +22,81 @@ import {
 } from "../commonElements"
 import { FormProps, InputValues, ModalType } from "../Types"
 import { MobileLoginValidator } from "../Validators"
+import { StepElement } from "Components/Wizard/types"
+
+interface LoginFormState {
+  advanceWizardStep: boolean
+  error: string
+  hideBackButton: boolean
+  steps: StepElement[]
+}
 
 class MobileLoginFormWithSystemContext extends Component<
   FormProps & {
     relayEnvironment: Environment
-  }
+  },
+  LoginFormState
 > {
+  static getDerivedStateFromProps(nextProps, prevState): LoginFormState | null {
+    if (nextProps.error !== prevState.error) {
+      if (nextProps.error === "missing two-factor authentication code") {
```

Comment:
This is brittle and it would be better for Gravity to return a stable error code in addition to the human-friendly error message. Doesn't have to be numeric. Could be `missing_otp` or `otp_required`.

Reply:
I wonder if we could replace the `:invalid_grant` error code that Gravity [currently creates](https://github.com/artsy/gravity/blob/ebe4ecfa22422ae10a4489ed7a2d11bfac67e33e/app/controllers/oauth_controller.rb#L147-L148) with `:otp_required` and `:otp_invalid` — assuming that the error code could even bubble down this far.

(The OAuth spec seems to stipulate that it's [permissible to create new codes](https://tools.ietf.org/html/rfc6749#section-8.5) subject to a few restrictions)

Reply:
For sure. Let's follow-up with this :+1: